virttest/iscsi.py

```diff
@@ -833,5 +834,170 @@ def create_iSCSI(params, root_dir=data_dir.get_tmp_dir()):
+                        ${lun02} : {"emulated_image": ..., ...},
+                    },
+                    ${target02}: {
+                        ${lun01} : {"emulated_image": ..., ...},
+                        ${lun02} : {"emulated_image": ..., ...},
+                    },
+                    ...
+                }
+              If there's only one lun without any special assigned requests,
+              lun should be set to "default_lun", for example:
+                targets_luns = { ${target01}: { "default_lun": {...}}}
+
+        :param params: parameters dict for iSCSI
+        :param root_dir: path for image
+        """
+        self._targets_luns = ast.literal_eval(params.get("targets_luns"))
+        # stores the iscsiLIO objects based on each targets
+        self._targets_mapping = {}
+        for target in self._targets_luns.keys():
+            single_target_param = params.copy()
+            for lun in self._targets_luns[target].keys():
+                single_target_param.update(self._targets_luns[target][lun])
+                single_target_param["target"] = target
+                single_target_param["iscsi_backend"] = "fileio"
+                self._targets_mapping[target] = IscsiLIO(single_target_param, root_dir)
+                if lun not in ("default_lun",):
+                    self._targets_mapping[target].luns = lun
+
+    def query_targets(self, emulated_image=None, lun=None):
+        """
+        Dynamically filter the targets from image name given and lun given.
+
+        :param emulated_image: the image name. If none, do NOT check emulated_image
+        :type emulated_image: string
+        :param lun: the lun. If none, do NOT check lun
+        :type lun: string
+
+        :return: the targets in list or []
+        :rtype: list
+        """
+        cmd = "targetcli ls /iscsi 1"
+        target_info = process.run(cmd).stdout_text
+        targets = re.findall(r"iqn[\.]\S+:\S+", target_info)
+
+        filtered_targets = []
+        for target in targets:
+            cmd = "targetcli ls /iscsi/%s/tpg1/luns" % target
+            luns_info = process.run(cmd).stdout_text
+            if (
+                (not emulated_image and not lun)
+                or (emulated_image and not lun and emulated_image + ")" in luns_info)
+                or (
+                    emulated_image
+                    and lun
+                    and emulated_image + ")" in luns_info
+                    and " " + lun + " " in luns_info
+                )
+            ):
+                filtered_targets.append(target)
+        return filtered_targets
+
+    def set_chap_auth_target(self, target=None):
+        """
+        set up authentication information for every single initiator,
+        which provides the capability to define common login information
+        for all Endpoints in a TPG
+
+        :param target: the target.
+        :type target: string
+        """
+        for key in self._targets_mapping.keys():
```

Comment:
targets = [target] if target else self._targets_mapping.keys()
for target in targets:

Then we don't make the 'if' check for each loop of for

Reply:
> targets = [target] if target else self._targets_mapping.keys() for target in targets:
> 
> Then we don't make the 'if' check for each loop of for

Handle one target for one call.